src/v/redpanda/application.h

```diff
@@ -256,6 +262,37 @@ class application {
 
     bool datalake_enabled();
 
+    // Stop the service.
+    // The method should be invoked in the ss::thread context.
+    template<class Service>
+    void stop_service(Service& s, ss::sstring name) {
+        // This watchdog is triggered after short period of time (30s). It
+        // adds message to the log that service is taking a long time to
+        // shutdown on INFO level.
+        ssx::watchdog short_wd(short_shutdown_warning_timeout, [this, name] {
+            vlog(
+              _log.info,
+              "Service {} is taking more than {} seconds to shut down.",
+              name,
+              std::chrono::duration_cast<std::chrono::seconds>(
+                short_shutdown_warning_timeout)
+                .count());
+        });
+        // This watchdog is triggered after long period of time. This indicates
+        // a bug (most likely).
+        ssx::watchdog long_wd(long_shutdown_warning_timeout, [this, name] {
+            vlog(
+              _log.info,
+              "Service {} is taking more than {} seconds to shut down!",
+              name,
+              std::chrono::duration_cast<std::chrono::seconds>(
+                long_shutdown_warning_timeout)
+                .count());
+        });
+        vlog(_log.info, "Stopping {}", name);
+        s.stop().get();
```

Comment:
Let's log a "Stopped {}" message too for completeness.

Reply:
you can infer this based on next service in the list

Reply:
> you can infer this based on next service in the list

but what about the very last one in the list?